File: desktop-mend/src/main/java/co/samco/mend4/desktop/commands/Clean.java
package co.samco.mend4.desktop.commands;

import co.samco.mend4.core.Settings;
import co.samco.mend4.core.exception.CorruptSettingsException;
import co.samco.mend4.desktop.core.I18N;
import co.samco.mend4.desktop.exception.SettingRequiredException;
import co.samco.mend4.desktop.helper.SettingsHelper;
import co.samco.mend4.desktop.helper.ShredHelper;
import co.samco.mend4.desktop.output.PrintStreamProvider;

import javax.inject.Inject;
import java.io.IOException;
import java.util.Arrays;
import java.util.List;
import java.util.function.Function;

public class Clean extends Command {
    public static final String COMMAND_NAME = "clean";

    private final PrintStreamProvider log;
    private final SettingsHelper settingsHelper;
    private final ShredHelper shredHelper;
    private final Settings settings;
    private final I18N strings;

    private final List<Function<List<String>, List<String>>> behaviourChain = Arrays.asList(
            a -> assertSettingsPresent(a),
            a -> performClean(a)
    );

    @Inject
    public Clean(I18N strings, PrintStreamProvider log, SettingsHelper settingsHelper, ShredHelper shredHelper,
                 Settings settings) {
        this.strings = strings;
        this.log = log;
        this.settingsHelper = settingsHelper;
        this.shredHelper = shredHelper;
        this.settings = settings;
    }

    protected List<String> assertSettingsPresent(List<String> args) {
        try {
            settingsHelper.assertRequiredSettingsExist(new Settings.Name[]{
                            Settings.Name.DECDIR, Settings.Name.SHREDCOMMAND},
                    COMMAND_NAME);
        } catch (IOException | SettingRequiredException e) {
            failWithMessage(log, e.getMessage());
            return null;
        }
        return args;
    }

    public List<String> performClean(List<String> args) {
        try {
            String decDir = settings.getValue(Settings.Name.DECDIR);
            shredHelper.shredFilesInDirectory(decDir);
            log.err().println(strings.get("Clean.cleanComplete"));
        } catch (CorruptSettingsException | IOException e) {
            failWithMessage(log, e.getMessage());
        }
        return null;
    }

    @Override
    public void execute(List<String> args) {
        executeBehaviourChain(behaviourChain, args);
    }

    @Override
    public String getUsageText() {
        return strings.getf("Clean.usage", COMMAND_NAME);
    }

    @Override
    public String getDescriptionText() {
        return strings.getf("Clean.description", Settings.Name.SHREDCOMMAND.toString());
    }

    @Override
    protected List<String> getCommandAliases() {
        return Arrays.asList(COMMAND_NAME);
    }

}

File: desktop-mend/src/main/java/co/samco/mend4/desktop/commands/Clean.kt
package co.samco.mend4.desktop.commands

import co.samco.mend4.core.exception.CorruptSettingsException
import co.samco.mend4.desktop.core.I18N
import co.samco.mend4.desktop.dao.SettingsDao
import co.samco.mend4.desktop.helper.SettingsHelper
import co.samco.mend4.desktop.helper.ShredHelper
import co.samco.mend4.desktop.output.PrintStreamProvider
import java.util.function.Function
import javax.inject.Inject

class Clean @Inject constructor(
    private val strings: I18N,
    private val log: PrintStreamProvider,
    private val settingsHelper: SettingsHelper,
    private val shredHelper: ShredHelper,
    private val settings: SettingsDao
) : CommandBase() {
    private val behaviourChain = listOf(
        Function { a: List<String> -> assertSettingsPresent(a) },
        Function { a: List<String> -> performClean(a) }
    )

    private fun assertSettingsPresent(args: List<String>): List<String>? {
        try {
            settingsHelper.assertRequiredSettingsExist(
                arrayOf(SettingsDao.DEC_DIR, SettingsDao.SHRED_COMMAND),
                COMMAND_NAME
            )
        } catch (t: Throwable) {
            failWithMessage(log, t.message)
            return null
        }
        return args
    }

    private fun performClean(args: List<String>): List<String>? {
        try {
            val decDir = settings.getValue(SettingsDao.DEC_DIR)
                ?: throw CorruptSettingsException(
                    strings.getf(
                        "General.dirRequired",
                        SettingsDao.DEC_DIR,
                        COMMAND_NAME
                    )
                )

            shredHelper.shredFilesInDirectory(decDir)
            log.err().println(strings["Clean.cleanComplete"])
        } catch (t: Throwable) {
            failWithMessage(log, t.message)
        }
        return null
    }

    public override fun execute(args: List<String>) {
        executeBehaviourChain(behaviourChain, args)
    }

    override val usageText: String
        get() = strings.getf("Clean.usage", COMMAND_NAME)
    override val descriptionText: String
        get() = strings.getf("Clean.description", SettingsDao.SHRED_COMMAND.toString())

    override val commandAliases: List<String>
        get() = listOf(COMMAND_NAME)

    companion object {
        const val COMMAND_NAME = "clean"
    }
}